src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPuller.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Azure.Messaging.EventGrid.Namespaces;
+using Microsoft.Extensions.Azure;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+
+namespace Workleap.DomainEventPropagation;
+
+internal class EventPuller : BackgroundService
+{
+    private readonly EventGridReceptionChannel[] _eventGridReceptionChannels;
+
+    public EventPuller(
+        IEnumerable<EventGridClientDescriptor> clientDescriptors,
+        IAzureClientFactory<EventGridClient> eventGridClientFactory,
+        IOptionsMonitor<EventPropagationSubscriptionOptions> optionsMonitor)
+    {
+        this._eventGridReceptionChannels = clientDescriptors.Select(descriptor =>
+            new EventGridReceptionChannel(
+                optionsMonitor.Get(descriptor.Name).TopicName,
+                optionsMonitor.Get(descriptor.Name).SubscriptionName,
+                eventGridClientFactory.CreateClient(descriptor.Name))).ToArray();
+    }
+
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        return Task.WhenAll(this._eventGridReceptionChannels.Select(channel => Task.Run(() => StartReceivingEventsAsync(channel, stoppingToken), stoppingToken)));
+    }
+
+    private static async Task StartReceivingEventsAsync(EventGridReceptionChannel channel, CancellationToken stoppingToken)
+    {
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            var result = await channel.Client.ReceiveCloudEventsAsync(channel.Topic, channel.Subscription, cancellationToken: stoppingToken).ConfigureAwait(false);
```

Comment:
We will probably need to `try/catch` this block so that neither the background service nor the application fails.